### DISCUSSION TITLE
Adding "GetDownloadFolder" to WebDriver class?

### DISCUSSION THREAD

**Discussion by GHRyunosuke** [GCuser99/SeleniumVBA#100] [Ideas]
While there is already "SetDownloadFolder", how about adding a corresponding "GetDownloadFolder" that returns the current configured DownloadFolder?

It could come in handy when using "WaitForDownload":

```
With driver
    .WaitForDownload .GetDownloadFolder & fileName
End with
```

**Comment by 6DiegoDiego9** [comment]
Hi @GHRyunosuke, how do you download the file?
If you use DownloadResource, the default path is the same, therefore you would not need to specify it again for WaitForDownload.

**Reply by GHRyunosuke** [reply]
I just click (via selenium VBA) the "download" button one would humanly click to download a file. Sometimes I would use "SetDownloadFolder" to set a specific folder, and then when i use "WaitForDownload", i kind of thinking, oh where did I set that folder to be previously ...

**Comment by GCuser99** [comment]
@GHRyunosuke I think that's an idea worth considering, but I can't find a straight-forward way of doing it. 

The ```SetDownloadFolder``` method uses the [Chrome Dev Protocol](https://chromedevtools.github.io/devtools-protocol/tot/Browser/#method-setDownloadBehavior), but I haven't yet found a corresponding "get" command available. In principle, one could scrape the "location" element for the url "chrome://settings/downloads", but I think that would be awkward. It might be possible to hack the Chrome/Edge User Data directories for the desired info but solving that is beyond my pay grade :-). If you (or anyone else) can find a clean way of getting the browser's default download directory directly let us know and we will consider...

Anyway, I think the factory setting default folder for MOST systems is "%USERPROFILE%\Downloads". But the user/admin can alter that and set it to a different location in user's Chrome/Edge profile, so it's not a given. 

The surest way to know what it is currently (in a programmatic sense) is to set it yourself. Do you use the SeleniumVBA.ini file? I use the [pre-loaded capabilities feature](https://github.com/GCuser99/SeleniumVBA/wiki#advanced-customization) in that to set the default download folder to my preference of ".\\" so that I do not have to worry about setting it at runtime. That way I always know where the default location is and I can change that at runtime if needed using ```SetDownloadFolder```.  If you are interested in going that route, let us know if you have any questions about it - would love to assist if needed.

Cheers, Mike

PS: you might be thinking that it would be feasible to save the argument value from ```SetDownloadFolder``` as a global in the WebDriver class and then simply retrieve it with ```GetDownloadFolder```. But that solution presupposes that ```SetDownloadFolder``` is always executed prior, which cannot be guaranteed (in my useage hardly ever). A reliable solution would return the default location directly from the browser itself.

**Reply by 6DiegoDiego9** [reply]
If we can get the path of the profile folder of Chrome, we could also get it from the Preferences file as documented [here](https://stackoverflow.com/a/18438722/11738627).

**Reply by GHRyunosuke** [reply]
Thank you @GCuser99 for the kind reply,  I just thought a "get" method would be easier than a "set" method... seems the background of this case is much deeper. 

I think I would just use a variable to store the folder path when I use "SetDownloadFolder " and reuse it in "WaitForDownload" later.

**Reply by GCuser99** [reply]
Yeah @6DiegoDiego9 I had tried searching that "Preferences" file before I wrote my reply. My Chrome "Preferences" file does not have the "default_directory" key in the "download" dictionary. Does your Preferences have it?

**Reply by 6DiegoDiego9** [reply]
I see this:

![image](https://github.com/GCuser99/SeleniumVBA/assets/109476941/eb5eb92a-5b34-4e54-81e5-6ff857f21b35)

**Reply by GCuser99** [reply]
Yeah, I have that key under "savefile" dictionary, but it points to some file I manually saved a long time ago, not the downloads directory. Thanks for looking into it @6DiegoDiego9.
